source/flux/parts/courses.js

```diff
@@ -68,23 +69,70 @@ export function updateCourseData(): UpdateCourseDataActionType {
 	}
 }
 
+const LOAD_RECENT_SEARCHES = 'courses/LOAD_RECENT_SEARCHES'
+
+type LoadRecentSearchesAction = {|
+	type: 'courses/LOAD_RECENT_SEARCHES',
+	payload: string[],
+|}
+export async function loadRecentSearches(): Promise<LoadRecentSearchesAction> {
+	const recentSearches = await storage.getRecentSearches()
+	return {type: LOAD_RECENT_SEARCHES, payload: recentSearches}
+}
+
+const UPDATE_RECENT_SEARCHES = 'courses/UPDATE_RECENT_SEARCHES'
+
+type UpdateRecentSearchesAction = {|
+	type: 'courses/UPDATE_RECENT_SEARCHES',
+	payload: string[],
+|}
+export function updateRecentSearches(
+	query: string,
+): ThunkAction<UpdateRecentSearchesAction> {
+	return (dispatch, getState) => {
+		const state = getState()
+
+		const oldRecentSearches = state.courses ? state.courses.recentSearches : []
+		if (oldRecentSearches.includes(query)) {
+			return
+		}
+		let newRecentSearches = oldRecentSearches
+		if (newRecentSearches.length === 0) {
+			newRecentSearches = [query]
+		} else {
+			newRecentSearches.unshift(query)
```

Comment:
newRecentSearches = [query, ...newRecentSearches]

I tend to avoid mutation methods except in performance sensitive code? You can push back if you wish.

Reply:
If I change it to `newRecentSearches = [query, ...newRecentSearches]` it seems to break the searching...as in when I press return after typing in a query, it doesn't refresh the search results unless I do it again. Works fine as is..maybe the ... is too slow?

Reply:
Meh. Ok.

Reply:
Actually it's working now..? May have just been slow memory on my computer...I keep getting notifications that my memory is maxed out..

Reply:
I would have a hard time believing that there was any performance difference between those two…